### PR TITLE
[no-release-notes] Add sequence_table testing

### PR DESCRIPTION
**Pull request by max-hoffman**

### CODE REVIEW

File: sql/plan/sequence_table.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+func (s SequenceTableFn) Children() []sql.Node {
+	return []sql.Node{}
+}
+
+func (s SequenceTableFn) RowIter(_ *sql.Context, _ sql.Row) (sql.RowIter, error) {
+	rowIter := &SequenceTableFnRowIter{i: 0, n: s.len}
+	return rowIter, nil
+}
+
+func (s SequenceTableFn) WithChildren(_ ...sql.Node) (sql.Node, error) {
```
Comment (PavelSafronov) [inline sql/plan/sequence_table.go:83]: This type doesn't really support children, so should we return an error here?

Same question for other methods like `Expressions`/`WithExpressions` and `Database`/`WithDatabase`.

**Review by zachmu — COMMENTED** [review]
You were right, I hate it. Don't do this.

If you want to test the use of table functions in planning, just use the method already set out by TestTableFunctions, it works fine for that purpose.

You could also just use JSONTable for some tests, I think

File: sql/plan/sequence_table.go (hunk 1 above)
Comment (max-hoffman) [reply sql/plan/sequence_table.go]: WithChildren and WithExpressions are called during our default DFS traversals. Having them no-op is desirable.

File: enginetest/engine_only_test.go [hunk 2]
```diff
@@ -650,15 +650,33 @@ func TestTableFunctions(t *testing.T) {
...
+			Expected: []sql.Row{{0}, {1}},
+		},
 	}
 
 	harness := enginetest.NewMemoryHarness("", 1, testNumPartitions, true, nil)
 	harness.Setup(setup.MydbData)
 
 	databaseProvider := harness.NewDatabaseProvider()
-	testDatabaseProvider := NewTestProvider(&databaseProvider, SimpleTableFunction{})
+	databaseProvider.(*memory.DbProvider).SetTableFunction("simple_table_function", SimpleTableFunction{})
```
Comment (zachmu) [inline enginetest/engine_only_test.go:675]: Why don't you use NewTestProvider as before? Just give it additional table functions to supply

File: enginetest/memory_harness.go [hunk 3]
```diff
@@ -232,11 +239,14 @@ func (m *MemoryHarness) getProvider() sql.MutableDatabaseProvider {
...
+	if m.provider != nil {
+		return m.provider
+	}
+
+	return memory.NewDBProviderWithOpts(
 		memory.NativeIndexProvider(m.nativeIndexSupport),
 		memory.HistoryProvider(true),
-	}
-	return memory.NewDBProviderWithOpts(opts...)
+		TableFunctionProvider("sequence_table", &memory.IntSequenceTable{}))
```
Comment (zachmu) [inline enginetest/memory_harness.go:249]: Shouldn't be needed, right?

File: sql/analyzer/apply_join.go [hunk 4]
```diff
@@ -76,7 +76,7 @@ func transformJoinApply(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope,
 					// which aren't possible when values are projected
 					// above join filter
 					rt := getResolvedTable(n)
-					if plan.IsDualTable(rt.Table) {
+					if rt != nil && plan.IsDualTable(rt.Table) {
```
Comment (zachmu) [inline sql/analyzer/apply_join.go:79]: Did you intend to include this?

File: enginetest/engine_only_test.go [hunk 5]
```diff
@@ -650,15 +650,33 @@ func TestTableFunctions(t *testing.T) {
...
+	databaseProvider.(*memory.DbProvider).SetTableFunction("simple_TABLE_function", SimpleTableFunction{})
```
Comment (zachmu) [inline enginetest/engine_only_test.go:676]: Curious why this is necessary. Aren't we already doing case-insensitive matching here?

File: enginetest/engine_only_test.go [hunk 6]
```diff
@@ -650,15 +650,33 @@ func TestTableFunctions(t *testing.T) {
 			Query:    "SELECT * from (select * from simple_TABLE_function(123)) as tf;",
 			Expected: []sql.Row{{"foo", 123}},
 		},
+		{
```
Comment (zachmu) [inline enginetest/engine_only_test.go:653]: Weren't there plan tests in the original PR as well? What happened to those?

**Review by zachmu — APPROVED** [review]
Tests and new test function are good, but it seems like the changes to the memory provider are unnecessary.

File: enginetest/engine_only_test.go (hunk 6 above)
Comment (max-hoffman) [reply enginetest/engine_only_test.go:653]: i don't like having disparate plan tests that have to be updated by hand. i can repro all of my errors with correctness tests so i'll stick with those for now.

**Pull request merged**